libdrgn/debug_info.c

```diff
@@ -2127,6 +2128,19 @@ drgn_type_from_dwarf_internal(struct drgn_debug_info *dbinfo, Dwarf_Die *die,
 					 "maximum DWARF type parsing depth exceeded");
 	}
 
+	// Follow signature tags on die
```

Comment:
Right now the positioning of this is a little off - If the signature points to a incomplete declaration, it won't be resolved like normal. To fix that, the simplest solution is to do this translation when lazy objects are created (as I think that's where the declarations are resolved). However, since that area of code is likely changing a lot, I wanted to avoid touching it as much as possible so I put this here.

Reply:
Ah, previously, the methods work added a find_definition as part of the lazy object port, as method members are frequently pointed to as incomplete types. Since (on that previous patch) this lookup is done on lazy type creation, it wasn't done in this case.

https://github.com/jgkamat/drgn/blob/07185513740e33d79d95420c26b6aad5de39c7c9/libdrgn/debug_info.c#L1249

 I'm not sure if this is an issue for templates on their own, but it is a concern here - especially on gcc, a signature tag often points to an incomplete type (which must use `find_definition` to resolve properly). However, right now this lookup (on master) only occurs on indexing, which leaves some types as incomplete. 

Would you prefer me to just use find_definition right here for now? I *think* this could also be moved to lazy type creation (along with a declaration lookup) - up to you.

Reply:
Even without type units, I think you're right that if `drgn_type_from_dwarf_internal()` gets a DIE with `DW_AT_declaration`, it should try to find the definition. Would you mind opening a separate PR with just "libdrgn: Switch to addresses for indexing into definition map", and a commit on top of that adding the appropriate `find_definition()` to `drgn_type_from_dwarf_internal()`? I can merge that right away while we figure out the remaining nuances of signatures.

There's a minor question of whether to follow the declaration to its definition before or after checking the DWARF type cache (`drgn_debug_info::types`). The cache is a mapping from DIE address to the parsed `drgn_type`. Should we check the cache for the declaration first, then follow the declaration, and cache the type both for the declaration and the definition? This is faster for subsequent accesses to the declaration, since we won't need to follow the definition again, but it adds more entries to the cache.

For this PR, I think we do want the `DW_AT_signature` lookup you're doing here.